### PR TITLE
Optional retention filter for DB

### PR DESCRIPTION
**Pull request by codesome**
Closes https://github.com/prometheus/prometheus/issues/7576

It is a very small refactoring which will help Cortex and Thanos.

/cc @bwplotka @brancz @pracucci

### CODE REVIEW

File: tsdb/db.go [hunk 1]
```diff
@@ -560,14 +566,18 @@ func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs
...
-		opts:        opts,
-		compactc:    make(chan struct{}, 1),
-		donec:       make(chan struct{}),
-		stopc:       make(chan struct{}),
-		autoCompact: true,
-		chunkPool:   chunkenc.NewPool(),
+		dir:             dir,
+		logger:          l,
+		opts:            opts,
+		compactc:        make(chan struct{}, 1),
```
Comment (bwplotka) [inline tsdb/db.go:572]: actually it's the time in my life I see `c` suffix fo channel :thinking:

File: tsdb/db.go [hunk 2]
```diff
@@ -118,6 +118,11 @@ type Options struct {
 	// SeriesLifecycleCallback specifies a list of callbacks that will be called during a lifecycle of a series.
 	// It is always a no-op in Prometheus and mainly meant for external users who import TSDB.
 	SeriesLifecycleCallback SeriesLifecycleCallback
+
+	// RetentionFilter is a function which returns the blocks which can be deleted.
+	// It is always the default time and size based retention in Prometheus and
+	// mainly meant for external users who import TSDB.
+	RetentionFilter func(blocks []*Block) map[ulid.ULID]*Block
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/db.go:125]: In that case, it could just return a slice of ULID :)

[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/db.go:125]: Or maybe map is better to avoid needed addition deduplication calculation

**Comment by roidelapluie** [comment]
Could we make our filters,e.g. beyondTimeRetention , public? does that make sense for thanos/cortex?

**Comment by codesome** [comment]
> Could we make our filters,e.g. beyondTimeRetention , public? does that make sense for downstreams?

I had made only the combination public. But it would make sense to make both public. I will do it in this PR.

**Comment by codesome** [comment]
I have updated the PR with
1. Name change
2. `*Block -> struct{}`
3. Additional `type BlocksToDeleteFunc func(blocks []*Block) map[ulid.ULID]struct{}` for a single place to tinker the function signature.

File: tsdb/db.go [hunk 3]
```diff
@@ -119,24 +119,26 @@ type Options struct {
 	// It is always a no-op in Prometheus and mainly meant for external users who import TSDB.
 	SeriesLifecycleCallback SeriesLifecycleCallback
 
-	// RetentionFilter is a function which returns the blocks which can be deleted.
+	// BlockToDelete is a function which returns the blocks which can be deleted.
 	// It is always the default time and size based retention in Prometheus and
 	// mainly meant for external users who import TSDB.
-	RetentionFilter func(blocks []*Block) map[ulid.ULID]*Block
+	BlockToDelete func(blocks []*Block) map[ulid.ULID]struct{}
```
Comment (pracucci) [inline tsdb/db.go:125]: Then why not `BlocksToDelete()` given it handle 1+ blocks? :)

File: tsdb/db.go [hunk 4]
```diff
@@ -118,19 +118,27 @@ type Options struct {
 	// SeriesLifecycleCallback specifies a list of callbacks that will be called during a lifecycle of a series.
 	// It is always a no-op in Prometheus and mainly meant for external users who import TSDB.
 	SeriesLifecycleCallback SeriesLifecycleCallback
+
+	// BlocksToDelete is a function which returns the blocks which can be deleted.
+	// It is always the default time and size based retention in Prometheus and
+	// mainly meant for external users who import TSDB.
+	BlocksToDelete func(blocks []*Block) map[ulid.ULID]struct{}
```
Comment (pracucci) [inline tsdb/db.go:125]:
```suggestion
	BlocksToDelete BlocksToDeleteFunc
```

Comment (codesome) [reply tsdb/db.go:125]: :facepalm: Thanks!

**Pull request merged**